File: 20_valid_parentheses/isValid.py
class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        parentheses = {
            ')': '(',
            '}': '{',
            ']': '[',
        }

        for i in s:
            if i in parentheses:
                if stack and stack[-1] == parentheses[i]:
                    stack.pop()
                else:
                    return False
            else:
                stack.append(i)
        
        if len(stack) > 0:
            return False

        return True
    
    

File: 20_valid_parentheses/is_valid.rs
use std::collections::HashMap;

impl Solution {
    pub fn is_valid(s: String) -> bool {
        let mut stack: Vec<char> = Vec::new();

        let mut p = HashMap::new();
        p.insert(')', '(');
        p.insert(']', '[');
        p.insert('}', '{');

        for c in s.chars() {
            if p.contains_key(&c) {
                if !stack.is_empty() && stack.last() == p.get(&c) {
                    stack.pop();
                } else {
                    return false;
                }
            } else {
                stack.push(c);
            }
        }

        if !stack.is_empty() {
            return false;
        }

        return true;
    }
}